test/utilities.jl

```diff
@@ -0,0 +1,108 @@
+
+    relative_diff = diff ./ abs.(reference)
+    max_rel_error_magnitude_loc = argmax(relative_diff)
+    max_rel_error_magnitude = relative_diff[max_rel_error_magnitude_loc]
+
+    lines = isnothing(err_msg) ? [] : [err_msg]
+    push!(lines, string("The arrays aren't consistent to within atol = ", atol,
+                        " rtol = ", rtol))
+
+    if (max_rel_error_magnitude == -1.0) && ((rtol != 0.0) || (atol == 0.0))
+        push!(lines, "Max Rel Diff Magnitude: ∞")
+    elseif ((rtol != 0.0) || (atol == 0.0)) && isnothing(max_rel_error_magnitude_loc)
+        push!(lines, string("Max Rel Diff Magnitude: ", max_rel_error_magnitude))
+    elseif ((rtol != 0.0) || (atol == 0.0))
+        push!(lines, string("Max Rel Diff Magnitude: ", max_rel_error_magnitude, " at ",
+                            error_location_fmt(max_rel_error_magnitude_loc)))
+    end
+
+    if ((atol != 0.0) || (rtol == 0.0)) && isnothing(max_abs_diff_loc)
+        push!(lines, string("Max Abs Diff Magnitude: ", max_abs_diff))
+    elseif ((atol != 0.0) || (rtol == 0.0))
+        push!(lines, string("Max Abs Diff Magnitude: ", max_abs_diff, " at ",
+                            error_location_fmt( max_abs_diff_loc)))
+    end
+
+    throw(ErrorException(join(lines, "\n")))
+    false
+end
+
+"""
+assert_allclose_grid(actual, reference, independent_vars; [rtol], [atol], [err_msg])
+
+This function is a convenience wrapper around assert_allclose that provides slightly nicer error 
+messages when actual and reference have the same shape. 
+
+`independent_vars` should be a `ndims(actual)` element array. The ith element of `independent_vars` 
+should be 2-tuple or 3-tuple:
+- `independent_vars[i][1]` should specify the name of the independent variable along the ith axis
+- `independent_vars[i][2]` should specify the values of the independent variable along the ith axis
+- `independent_vars[i][3]`, if present, should specify the units of the independent variable.
+"""
+function assert_allclose_grid(actual, reference, independent_vars; kwargs...)
+    @assert length(independent_vars) == ndims(actual)
+
+    function error_location_fmt(index)
+        tmp = []
+        for ax in 1:ndims(actual)
+            name, vals, units = if length(independent_vars[ax]) == 2
+                _name, _vals = independent_vars[ax]
+                _name, _vals, nothing
+            elseif length(independent_vars[ax]) == 3
+                independent_vars[ax]
+            else
+                throw(string("independent_vars[",ax,"] should only have 2 or 3 elements"))
+            end
+            if length(vals) != size(actual)[ax]
+                throw(string("length(independent_vars[",ax,"][2]) should be ", length(vals)))
+            end
+
+            if isnothing(units)
+                push!(tmp, string(name, " = ", vals[index[ax]]))
+            else
+                push!(tmp, string(name, " = ", vals[index[ax]], " ", units))
+            end
+        end
```

Comment:
Can I make two requests?  The first is that you rewrite the loop into a map, which I think is easier to reasonable once you are used to it.  The second is that you rename `tmp` to something more meaningful.  Maybe `ax_strs`?  I'll leave it up to you whether you want to adopt the code golf demonstrated below to move the conditional inside the arguments of `string`.

```julia
ax_strs = map(1:ndims(actual)) do ax 

    ...

    string(name, " = ", vals[index[ax]], (isnothing(units) ? "" : " $units"))
end
```